### PR TITLE
Add gradient accumulation support for distribution strategy

### PR DESCRIPTION
**Pull request by fanshiqing**
PR for issue #32176 .

There are still some issues to be fixed or further discussed:
(1）Gradient accumulation will break the default behavior of `basic_session_run_hooks`, e.g.:
```
I0917 17:51:10.444422 139809108260672 basic_session_run_hooks.py:262] loss = 0.435526, step = 0
I0917 17:51:10.495073 139809108260672 basic_session_run_hooks.py:692] global_step/sec: 19.6194
W0917 17:51:10.495340 139809108260672 basic_session_run_hooks.py:724] It seems that global step (tf.train.get_global_step) has not been increased. Current value (could be stable): 0 vs previous value: 0. You could increase the global step by passing tf.train.get_global_step() to Optimizer.apply_gradients or Optimizer.minimize.
I0917 17:51:10.495481 139809108260672 basic_session_run_hooks.py:260] loss = 0.435526, step = 0 (0.051 sec)
I0917 17:51:10.496573 139809108260672 basic_session_run_hooks.py:260] loss = 0.017421033, step = 1 (0.001 sec)
I0917 17:51:10.497781 139809108260672 basic_session_run_hooks.py:692] global_step/sec: 362.328
W0917 17:51:10.498049 139809108260672 basic_session_run_hooks.py:724] It seems that global step (tf.train.get_global_step) has not been increased. Current value (could be stable): 1 vs previous value: 1. You could increase the global step by passing tf.train.get_global_step() to Optimizer.apply_gradients or Optimizer.minimize.
I0917 17:51:10.498194 139809108260672 basic_session_run_hooks.py:260] loss = 0.017421033, step = 1 (0.002 sec)
I0917 17:51:10.499253 139809108260672 basic_session_run_hooks.py:260] loss = 0.00069683883, step = 2 (0.001 sec)
I0917 17:51:10.506485 139809108260672 basic_session_run_hooks.py:606] Saving checkpoints for 2 into /tmp/tmpYRa9ZY/model.ckpt.
```
`config.log_step_count_steps` should be scaled `iter_size` times larger.

(2) Currently gradients are *accumulated* locally, which implies that learning rate is adjusted `iter_size` times larger.

(3) This PR only support gradient accumulation for distribution strategy currently. Support for non-distribution case is natural as this GA realization is independent of optimizer and can be reused easily.

(4) GA add one new parameter `iter_size` to base class `Optimizer`, whether this will break some default behavior of some optimizer wrapper?

(5) Higher test coverage.

### CODE REVIEW

**Comment by yangjunpro** [comment]
@yuefengz 
Could you please help take a look at this PR? 

Thanks

**Comment by ustcwlin** [comment]
Like to hear why " it may not worth the time to add new features to Estimator now."

**Comment by yuefengz** [comment]
Thank you for your PR! I believe this is an important feature to have as well.

I have a couple suggestions out of the concern that we would like to make our API stable. It looks to me that the logic is not deeply coupled with an optimizer. Is it possible for you to put GA in a separate wrapper optimizer class? Also, since we are promoting TF 2.0, it would be more future-proof to make the wrapper a subclass of [optimizer v2](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/optimizer_v2/optimizer_v2.py) and make it eager-compatible.

**Comment by fanshiqing** [comment]
@yuefengz Okay, I will try to put GA in a separate wrapper optimizer class.

**Comment by yuefengz** [comment]
Thanks! I also believe this looks perfect as an [addon](https://github.com/tensorflow/addons/blob/master/CONTRIBUTING.md).  cc @facaiy

**Comment by facaiy** [comment]
@yuefengz Thanks for ping me, Yuefeng. My first thought: I agree with @ppwwyyxx that we'd better to implement GA as an optimizer wrapper. Please refer to:
+ https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/moving_average.py
+ https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/weight_decay_optimizers.py

GA looks like a useful feature for community, what do you think, @seanpmorgan @windqaq @squadrick?

**Comment by seanpmorgan** [comment]
> @yuefengz Thanks for ping me, Yuefeng. My first thought: I agree with @ppwwyyxx that we'd better to implement GA as an optimizer wrapper. Please refer to:
> 
> * https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/moving_average.py
> * https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/weight_decay_optimizers.py
> 
> GA looks like a useful feature for community, what do you think, @seanpmorgan @WindQAQ @Squadrick?

Agree and would be happy to have it as part of Addons. As just another example of optimizer wrapper that we currently have there is lookahead:
https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/lookahead.py

**Comment by facaiy** [comment]
@fanshiqing Shiqing, would you like to close this and open a PR to [addons](https://github.com/tensorflow/addons)?

**Comment by fanshiqing** [comment]
@facaiy Okay, I will close this one and open a new one to tf add-ons soon.

**Comment by andompesta** [comment]
I'm quite interested in it. Do you have already a working implementation ?

**Comment by atreyasha** [comment]
Similar idea is implemented here (https://github.com/run-ai/runai/blob/master/runai/ga/README.md) but unfortunately does not work with tensorflow.keras

**Comment by fernandocamargoti** [comment]
@fanshiqing, any update about it? I looked at tensorflow-addons and there's nothing there yet.
